### ISSUE TITLE
VS Code navigation steps

### ISSUE THREAD

**Issue by alexweininger** [vsls-contrib/codetour#74]
Currently CodeTour is an awesome way to let users interact with files and the text within them.

However, I believe that many projects would benefit from steps that could lead users through the VS Code UI in addition to the project files.

Examples include:
- Directing the user to VS Code's SCM view, and helping the user commit and push code
- Using a CodeTour to create a tutorial for a VS Code extension, a good example might be the Azure Functions extension. Where we could show the user how to create and deploy an Azure Function.

I think that implementing this feature could be as simple as just letting us define custom CodeTour steps, that have a VS Code command associated with them.

**Comment by lostintangent** [comment]
Hey! 👋 I’ve actually been meaning to add support for steps to include an optional list of commands. That way, authors could automate arbitrarily rich behavior, in addition to being able to easily associate content with a file or directory.

In your case, it sounds like you’d want to be able to define [content steps](https://github.com/vsls-contrib/codetour/blob/master/README.md#content-steps) (I.e. steps that aren’t associated with a specific file or folder), that also include a set of commands, that would be run when that step is played back?

I wonder if in addition to being able to specify a list of commands, we shouldn’t also add first-class support for a step that focuses a specific view (e.g. a “view step”). That way, you’d only need to rely on the custom command solution, if your step needed to do more than focus a view and provide descriptive text for it.

**Comment by alexweininger** [comment]
The "view step" would be exactly what we are looking for.

On top of a bare bones "view step" it might be useful to be able to highlight specific buttons, or text fields of the views. This could definitely be a stretch goal, because just providing a "view step" would be a huge help for us.

As you may know @fiveisprime and I are working on a trial app experience for the Azure App Service VS Code extension. And we think CodeTour could be perfect for a tutorial that could lead users through creating, and editing trial apps in VS Code.

**Comment by lostintangent** [comment]
@alexweininger Hey! I just pushed experimental support for two things: associating a view with a step, and adding an array of commands to a step. Both of these (currently) require you to hand edit the tour JSON file currently, since I want to get some feedback from you before I try to smooth over the authoring experience :)

To use this, you can create a new tour, add a content step (see image below) with the content you want, and then open up the tour file add a "view" and/or "commands" property. The "view" property simply refers to the ID of the VS Code view (e.g. `gistpad.gists`). When this property is set, and the user navigates to this step, CodeTour will automatically focus this view.

Additionally, the `commands` property is an array of strings, which represent the ID of VS Code commands. When the step is navigated to, CodeTour will execute those commands in the order specified. If you need to pass arguments, simply add a query string to the command name, that includes an array of arguments (e.g. `codetour.startTour?["Title"]`).  I'd love to hear any feedback/questions to know if this generally seems like it would work for your scenarios. Thanks!

![image](https://user-images.githubusercontent.com/116461/83437093-fde14200-a3f3-11ea-81ab-6914b49e9e87.png)

**Comment by alexweininger** [comment]
@lostintangent Thanks for making these changes so quickly!

Using the View IDs found here [here](https://code.visualstudio.com/docs/getstarted/keybindings#_activefocused-view-or-panel-when-clause-context) I'm getting the following error:

![image](https://user-images.githubusercontent.com/12476526/83442158-91b70c00-a3fc-11ea-98d3-01266aa0f5fe.png)

Here is the JSON:

![image](https://user-images.githubusercontent.com/12476526/83442191-a09dbe80-a3fc-11ea-8747-492d98ad4f8b.png)

**Comment by lostintangent** [comment]
Apologies! The change I checked in only supported 3rd-party views :/ If you grab v0.0.32, and set the `view` property to `scm`, it should work as expected. I added the notion of "well-known" views, that include all of the built-in ones (e.g. debug, explorer, search, scm).

**Comment by alexweininger** [comment]
The `view` property works really well and is exactly what we need. Thanks!

**Comment by lostintangent** [comment]
@alexweininger Awesome! 🙌I'll go ahead and close this issue then, since the `commands` list is also available for you if you need it. Thanks said, please file any other issues/suggestions on how we could unblock your use case.

In the future, I'm going to see how I can add a way to create "view steps" without manually editing the tour file, but I think I'll need some support from VS Code for that 👍
